File: NET_8/FoodCall.Domain/Entities/Review.cs
using FoodCall.Domain.Exceptions;

namespace FoodCall.Domain.Entities;

public class Review
{
    public Guid Id { get; private set; }
    public Guid OrderId { get; private set; }
    public Guid CustomerId { get; private set; }
    
    // Avaliação do Restaurante
    public int RestaurantRating { get; private set; } // 1 a 5
    public string? RestaurantComment { get; private set; }
    
    // Avaliação do Entregador
    public int? CourierRating { get; private set; } // Opcional (caso seja retirada no local)
    public string? CourierComment { get; private set; }
    
    public DateTime CreatedAt { get; private set; }

    public Review(Guid orderId, Guid customerId, int restaurantRating, string? restaurantComment)
    {
        ValidateOrderId(orderId);
        ValidateCustomerId(customerId);
        ValidateRating(restaurantRating);
        ValidateComment(restaurantComment, "Restaurant");

        Id = Guid.NewGuid();
        OrderId = orderId;
        CustomerId = customerId;
        RestaurantRating = restaurantRating;
        RestaurantComment = restaurantComment;
        CreatedAt = DateTime.UtcNow;
    }

    public void AddCourierReview(int rating, string? comment)
    {
        if (CourierRating.HasValue)
            throw new BusinessRuleException("AddCourierReview", "Avaliação do entregador já foi registrada");

        ValidateRating(rating);
        ValidateComment(comment, "Courier");

        CourierRating = rating;
        CourierComment = comment;
    }

    public void UpdateRestaurantReview(int rating, string? comment)
    {
        ValidateRating(rating);
        ValidateComment(comment, "Restaurant");

        RestaurantRating = rating;
        RestaurantComment = comment;
    }

    public void UpdateCourierReview(int rating, string? comment)
    {
        if (!CourierRating.HasValue)
            throw new BusinessRuleException("UpdateCourierReview", "Não há avaliação do entregador para atualizar");

        ValidateRating(rating);
        ValidateComment(comment, "Courier");

        CourierRating = rating;
        CourierComment = comment;
    }

    private void ValidateOrderId(Guid orderId)
    {
        if (orderId == Guid.Empty)
            throw new InvalidEntityException("Review", "OrderId não pode ser vazio");
    }

    private void ValidateCustomerId(Guid customerId)
    {
        if (customerId == Guid.Empty)
            throw new InvalidEntityException("Review", "CustomerId não pode ser vazio");
    }

    private void ValidateRating(int rating)
    {
        if (rating < 1 || rating > 5)
            throw new InvalidEntityException("Review", "A nota deve ser entre 1 e 5");
    }

    private void ValidateComment(string? comment, string entity)
    {
        if (!string.IsNullOrWhiteSpace(comment) && comment.Length > 500)
            throw new InvalidEntityException("Review", $"Comentário do {entity} deve ter no máximo 500 caracteres");
    }
}

File: Java/src/main/java/com/foodcall/domain/entities/Review.java
package com.foodcall.domain.entities;

import com.foodcall.domain.exceptions.BusinessRuleException;
import com.foodcall.domain.exceptions.InvalidEntityException;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "reviews")
@Getter
@NoArgsConstructor
public class Review {
    
    @Id
    private UUID id;
    
    @Column(nullable = false)
    private UUID orderId;
    
    @Column(nullable = false)
    private UUID customerId;
    
    @Column(nullable = false)
    private int restaurantRating; // 1 a 5
    
    @Column(length = 500)
    private String restaurantComment;
    
    private Integer courierRating; // Opcional
    
    @Column(length = 500)
    private String courierComment;
    
    @Column(nullable = false)
    private LocalDateTime createdAt;
    
    public Review(UUID orderId, UUID customerId, int restaurantRating, String restaurantComment) {
        validateOrderId(orderId);
        validateCustomerId(customerId);
        validateRating(restaurantRating);
        validateComment(restaurantComment, "Restaurant");
        
        this.id = UUID.randomUUID();
        this.orderId = orderId;
        this.customerId = customerId;
        this.restaurantRating = restaurantRating;
        this.restaurantComment = restaurantComment;
        this.createdAt = LocalDateTime.now();
    }
    
    public void addCourierReview(int rating, String comment) {
        if (courierRating != null) {
            throw new BusinessRuleException("AddCourierReview", "Avaliação do entregador já foi registrada");
        }
        
        validateRating(rating);
        validateComment(comment, "Courier");
        
        this.courierRating = rating;
        this.courierComment = comment;
    }
    
    public void updateRestaurantReview(int rating, String comment) {
        validateRating(rating);
        validateComment(comment, "Restaurant");
        
        this.restaurantRating = rating;
        this.restaurantComment = comment;
    }
    
    public void updateCourierReview(int rating, String comment) {
        if (courierRating == null) {
            throw new BusinessRuleException("UpdateCourierReview", "Não há avaliação do entregador para atualizar");
        }
        
        validateRating(rating);
        validateComment(comment, "Courier");
        
        this.courierRating = rating;
        this.courierComment = comment;
    }
    
    private void validateOrderId(UUID orderId) {
        if (orderId == null) {
            throw new InvalidEntityException("Review", "OrderId não pode ser vazio");
        }
    }
    
    private void validateCustomerId(UUID customerId) {
        if (customerId == null) {
            throw new InvalidEntityException("Review", "CustomerId não pode ser vazio");
        }
    }
    
    private void validateRating(int rating) {
        if (rating < 1 || rating > 5) {
            throw new InvalidEntityException("Review", "A nota deve ser entre 1 e 5");
        }
    }
    
    private void validateComment(String comment, String entity) {
        if (comment != null && !comment.trim().isEmpty() && comment.length() > 500) {
            throw new InvalidEntityException("Review", 
                String.format("Comentário do %s deve ter no máximo 500 caracteres", entity));
        }
    }
}